File: 129. Sum Root to Leaf Numbers/sumNumbers.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {number}
 */
const sumNumbers = (root) => {
  let res = 0;
  const dfs = (node, curr) => {
    if (!node) {
      return;
    }

    let num = 10 * curr + node.val;
    if (!node.left && !node.right) {
      res += num;
    }

    if (node.left) {
      dfs(node.left, num);
    }

    if (node.right) {
      dfs(node.right, num);
    }
  };
  
  dfs(root, 0);
  return res;
};


const root = new TreeNode(1);
root.left = new TreeNode(2);
root.right = new TreeNode(3);

console.log('sumNumbers: ', sumNumbers(root));

File: 129. Sum Root to Leaf Numbers/sumNumbers.py
from typing import  Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def sumNumbers(self, root: Optional[TreeNode]) -> int:
    res = 0
    def dfs(node, curr):
      nonlocal res
      if not node:
        return
      
      num = curr * 10 + node.val
      if not node.left and not node.right:
        res += num
        return
      
      if node.left:
        dfs(node.left, num)
        
      if node.right:
        dfs(node.right, num)
        
    dfs(root, 0)
    return res


root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

solution = Solution()
print('sumNumbers: ', solution.sumNumbers(root))
        
    
        
        